### PR TITLE
[CFM-835] control group request

### PR DESCRIPTION
**Pull request by anvmn**
https://issuetracker.amplexor.com/jira/browse/CFM-835

## Dependent on #1298

### CODE REVIEW

**Comment by anvmn** [comment]
Log and email, when group is approved:

out


Log, when group is rejected:

**Comment by anvmn** [comment]
out

File: project/profiles/capacity4more/modules/c4m/og/c4m_og/c4m_og.module [hunk 1]
```diff
@@ -185,10 +185,185 @@ function c4m_og_menu() {
     'access arguments' => array(1, 2),
   );
 
+  // Menu item for approval of group request url.
+  $items['group/%/approve'] = array(
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/og/c4m_og/c4m_og.module]: use `%node`

Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/og/c4m_og/c4m_og.module]: or maybe try to replace `'group'` => `'node'`

File: project/profiles/capacity4more/modules/c4m/og/c4m_og/c4m_og.module [hunk 2]
```diff
@@ -185,10 +185,185 @@ function c4m_og_menu() {
...
+ * Site administrators can access, as long as $gid is OG group.
+ *
+ * @param int $gid
+ *   Group ID.
+ *
+ * @return bool
+ *   Wither access is allowed.
+ */
+function c4m_og_control_group_request_access($gid) {
+  if (!user_access('administer group')) {
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/og/c4m_og/c4m_og.module]: `c4m_user_is_site_admin()`

File: project/profiles/capacity4more/modules/c4m/og/c4m_og/c4m_og.module [hunk 3]
```diff
@@ -185,10 +185,185 @@ function c4m_og_menu() {
...
+    return FALSE;
+  }
+
+  $node = node_load($gid);
+
+  return ($node->type == 'group') || ($node->type == 'project');
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/og/c4m_og/c4m_og.module]:
```php
return og_is_group('node', $group);
```

File: project/profiles/capacity4more/modules/c4m/og/c4m_og/c4m_og.module [hunk 4]
```diff
@@ -185,10 +185,185 @@ function c4m_og_menu() {
...
+ *
+ * Site administrators can access, as long as $gid is OG group.
+ *
+ * @param int $gid
+ *   Group ID.
+ *
+ * @return bool
+ *   Wither access is allowed.
+ */
+function c4m_og_control_group_request_access($gid) {
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/og/c4m_og/c4m_og.module]: `$gid` => `$group`

File: project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/c4m_user_notifications.module [hunk 5]
```diff
@@ -329,6 +329,29 @@ function c4m_user_notifications_form_c4m_og_reject_membership_request_alter(&$fo
...
+  $form['#submit'][] = 'c4m_og_route_to_group';
+}
+
+/**
+ * Implements hook_form_FORM_ID_alter().
+ *
+ * Triggers email notification when site administrator rejects group
+ * request.
+ */
+function c4m_user_notifications_form_c4m_og_reject_group_request_alter(&$form, &$form_state) {
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/c4m_user_notifications.module]: remove this and move the logic to the `hook_form`

File: project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/c4m_user_notifications.module [hunk 6]
```diff
@@ -329,6 +329,29 @@ function c4m_user_notifications_form_c4m_og_reject_membership_request_alter(&$fo
 }
 
 /**
+ * Implements hook_form_FORM_ID_alter().
+ *
+ * Triggers email notification when site administrator approves group
+ * request.
+ */
+function c4m_user_notifications_form_c4m_og_approve_group_request_alter(&$form, &$form_state) {
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/c4m_user_notifications.module]: remove this and move the logic to the `hook_form`

File: project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/includes/c4m_user_notifications_handlers.inc [hunk 7]
```diff
@@ -154,6 +154,29 @@ function c4m_user_notifications_moderated_group_notify_user_membership_rejected(
...
 
 /**
+ * Email notification to group owner, when group is approved.
+ */
+function c4m_user_notifications_notify_user_group_approved(&$form, &$form_state) {
+  $gid = $form_state['values']['gid'];
+  $admin_comment = $form_state['values']['message_from_admin'];
+  $group = node_load($gid);
+
+  if (!($message = message_create('c4m_group_request_approved', array('uid' => $group->uid)))) {
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/includes/c4m_user_notifications_handlers.inc]: remove the validation

File: project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/includes/c4m_user_notifications_handlers.inc [hunk 8]
```diff
@@ -819,3 +842,64 @@ function c4m_user_notifications_message_subscribe_get_subscribers($message, $sub
...
+ * @param object $form
+ *   Form array to populate.
+ * @param object $form_state
+ *   Form state array.
+ */
+function c4m_user_notifications_log_group_approved($form, $form_state) {
+  $gid = $form_state['values']['gid'];
+  $admin_comment = $form_state['values']['message_from_admin'];
+
+  _c4m_user_notifications_log_group_state_change('pending', 'deleted', $gid, $admin_comment);
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/includes/c4m_user_notifications_handlers.inc]: should be approved

**Comment by anvmn** [comment]
CFM-838:

out

**Comment by anvmn** [comment]
CFM-840:

out

**Comment by anvmn** [comment]
CFM-842:

out

**Comment by anvmn** [comment]
CFM-844:

selection_069

File: project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.features.inc [hunk 9]
```diff
@@ -780,6 +780,33 @@ function c4m_message_default_message_type() {
     },
     "rdf_mapping" : []
   }');
+  $items['c4m_notify_editors_content_updated'] = entity_import('message_type', '{
+    "name" : "c4m_notify_editors_content_updated",
+    "description" : "c4m Notify Editors Content Updated",
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.features.inc]: Please update the MT number in here,  thanks

File: project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.features.inc [hunk 10]
```diff
@@ -25,6 +25,33 @@ function c4m_message_views_api($module = NULL, $api = NULL) {
  */
 function c4m_message_default_message_type() {
   $items = array();
+  $items['c4m_content_updated'] = entity_import('message_type', '{
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.features.inc]: => `c4m_group_content_update`

File: project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.features.inc [hunk 11]
```diff
@@ -704,6 +780,33 @@ function c4m_message_default_message_type() {
     },
     "rdf_mapping" : []
   }');
+  $items['c4m_notify_editors_content_updated'] = entity_import('message_type', '{
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.features.inc]: => `c4m_notify_editors_group_content_updated`

File: project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module [hunk 12]
```diff
@@ -1744,6 +1814,19 @@ function c4m_message_format_help_section_url(Message $message) {
...
+ * Returns site contact link token.
+ *
+ * @param Message $message
+ *   The Message entity.
+ *
+ * @return string
+ *   Site contact link.
+ */
+function c4m_message_format_site_contact_link(Message $message) {
+  return 'mailto:capacity4dev@ec.europa.eu';
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/message/c4m_message/c4m_message.module]: @soniCaH 
can we use `variable_get('site_mail');` in here ?
or should we add another variable, something like `c4m_notifications_site_contant_mail`

File: project/profiles/capacity4more/modules/c4m/og/c4m_og/c4m_og.module [hunk 13]
```diff
@@ -185,10 +185,261 @@ function c4m_og_menu() {
     'access arguments' => array(1, 2),
   );
 
+  // Menu item for approval of group request url.
+  $items['node/%node/approve/group'] = array(
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/og/c4m_og/c4m_og.module]: => `node/%node/approve-group`

File: project/profiles/capacity4more/modules/c4m/og/c4m_og/c4m_og.module [hunk 14]
```diff
@@ -185,10 +185,261 @@ function c4m_og_menu() {
...
+    'page callback' => 'drupal_get_form',
+    'page arguments' => array('c4m_og_approve_group_request', 1),
+    'access callback' => 'c4m_og_control_group_request_access',
+    'access arguments' => array(1),
+  );
+
+  // Menu item for rejection of group request url.
+  $items['node/%node/reject/group'] = array(
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/og/c4m_og/c4m_og.module]: => `node/%node/reject-group`

File: project/profiles/capacity4more/modules/c4m/og/c4m_og/c4m_og.module [hunk 15]
```diff
@@ -185,10 +185,261 @@ function c4m_og_menu() {
...
+    'access arguments' => array(1),
+  );
+
   return $items;
 }
 
 /**
+ * Access to control group requests.
+ *
+ * Site administrators can access, as long as $gid is OG group.
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/og/c4m_og/c4m_og.module]: fix the docs.

File: project/profiles/capacity4more/modules/c4m/user/c4m_user/c4m_user.module [hunk 16]
```diff
@@ -1381,3 +1381,25 @@ function c4m_user_registrationpassword_confirmation_url($account) {
...
+ * Checks if user is site administrator.
+ *
+ * @param mixed $user
+ *   The user object. If not passed, current user is used.
+ *
+ * @return bool
+ *   Whether user is site administrator.
+ */
+function c4m_user_is_site_admin($user = NULL) {
+  if (!isset($user)) {
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user/c4m_user.module]: => `empty()`

File: project/profiles/capacity4more/modules/c4m/user/c4m_user/c4m_user.module [hunk 17]
```diff
@@ -1381,3 +1381,25 @@ function c4m_user_registrationpassword_confirmation_url($account) {
...
+    global $user;
+  }
+
+  // The user has the sitewide role 'administrator'.
+  if (in_array('administrator', $user->roles)) {
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user/c4m_user.module]: use `user_has_role`

File: project/profiles/capacity4more/modules/c4m/og/c4m_og/c4m_og.module [hunk 18]
```diff
@@ -185,10 +185,261 @@ function c4m_og_menu() {
...
+ *
+ * @param object $form
+ *   Form array to populate.
+ * @param object $form_state
+ *   Form state array.
+ * @param object $node
+ *   Group node.
+ *
+ * @return object
+ *   Approve group request form.
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/og/c4m_og/c4m_og.module]: remove the "request"

File: project/profiles/capacity4more/modules/c4m/og/c4m_og/c4m_og.module [hunk 19]
```diff
@@ -185,10 +185,261 @@ function c4m_og_menu() {
...
+ *   Use 'reject' when group is rejected.
+ * @param int $gid
+ *   Group ID.
+ */
+function c4m_og_control_group_request($action, $gid) {
+  // Check that group is at initial ('pending') state.
+  $wrapper = entity_metadata_wrapper('node', $gid);
+
+  if (!_c4m_og_group_at_initial_state($gid)) {
+    // If group is not at initial state, present warning message and return.
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/og/c4m_og/c4m_og.module]: replace the technical comment with a simple one, something like "request already handled ..."

File: project/profiles/capacity4more/modules/c4m/og/c4m_og/c4m_og.module [hunk 20]
```diff
@@ -185,10 +185,261 @@ function c4m_og_menu() {
...
+    $message = _c4m_og_format_group_request_failure_reason_message($gid);
+
+    drupal_set_message($message, 'warning');
+    return;
+  }
+
+  // Change group state and set message.
+  $bundle = $wrapper->getBundle();
+  switch ($action) {
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/og/c4m_og/c4m_og.module]:
```php
$message = t('@bundle request @action', array(
  '@bundle' => ucfirst($wrapper->getBundle()),
  '@action' => $action,
));
```

Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/og/c4m_og/c4m_og.module]:
```php
$a =array(
'approve' => 'draft',
'reject' => 'deleted',
);

$wrapper=>c4m_og_status->set($a[$action]);
```

File: project/profiles/capacity4more/modules/c4m/og/c4m_og/c4m_og.module [hunk 21]
```diff
@@ -185,10 +185,261 @@ function c4m_og_menu() {
...
+ * @return bool
+ *   Failure reason message if for invalid group request.
+ */
+function _c4m_og_format_group_request_failure_reason_message($gid) {
+  $state = _c4m_og_get_group_state($gid);
+
+  $wrapper = entity_metadata_wrapper('node', $gid);
+  $bundle = $wrapper->getBundle();
+
+  switch ($state) {
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/og/c4m_og/c4m_og.module]: fix the switch

File: project/profiles/capacity4more/modules/c4m/og/c4m_og/c4m_og.module [hunk 22]
```diff
@@ -302,6 +553,20 @@ function c4m_og_route_to_group_people($form, &$form_state) {
 }
 
 /**
+ * Forwards the page to group's People page.
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/og/c4m_og/c4m_og.module]: fix the docs

Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/og/c4m_og/c4m_og.module]: group admin page

File: project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/includes/c4m_user_notifications_handlers.inc [hunk 23]
```diff
@@ -153,6 +142,26 @@ function c4m_user_notifications_moderated_group_notify_user_membership_rejected(
...
 
 /**
+ * Email notification to group owner, when group is approved.
+ */
+function c4m_user_notifications_notify_user_group_approved(&$form, &$form_state) {
+  $gid = $form_state['values']['gid'];
+  $admin_comment = $form_state['values']['message_from_admin'];
+  $group = node_load($gid);
+
+  $message = message_create('c4m_group_request_approved', array('uid' => $group->uid));
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/includes/c4m_user_notifications_handlers.inc]: create the message on an external function and use it also on the `hook_node_update()`

File: project/profiles/capacity4more/modules/c4m/og/c4m_og/c4m_og.module [hunk 24]
```diff
@@ -421,11 +440,11 @@ function _c4m_og_format_group_request_failure_reason_message($gid) {
 
   $wrapper = entity_metadata_wrapper('node', $gid);
 
-  $message_states = array (
+  $message_states = array(
     'draft' => 'approved',
     'deleted' => 'rejected',
   );
-  $message_action = in_array($state, $message_states) ? $message_states[$state] : 'handled';
+  $message_action = in_array($state, array_keys($message_states)) ? $message_states[$state] : 'handled';
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/og/c4m_og/c4m_og.module]: either rejected or approved

File: project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/c4m_user_notifications.module [hunk 25]
```diff
@@ -443,10 +444,14 @@ function _c4m_user_notifications_process_general_triggers_update($node) {
         break;
 
       // 'published' -> 'draft' => Unpublished.
+      // 'pending' -> 'draft' => Group approved.
       case 'draft':
         if ($group_former_status == 'published') {
           c4m_user_notifications_group_status_set_unpublished($node, $group_former_status);
         }
+        elseif ($group_former_status == 'pending' && !isset($node->abort_notification)) {
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user_notifications/c4m_user_notifications.module]: use `empty()`

File: project/profiles/capacity4more/modules/c4m/og/c4m_og/c4m_og.module [hunk 26]
```diff
@@ -437,18 +437,11 @@ function _c4m_og_group_at_initial_state($gid) {
...
-  $message_states = array(
-    'draft' => 'approved',
-    'deleted' => 'rejected',
-  );
-  $message_action = in_array($state, array_keys($message_states)) ? $message_states[$state] : 'handled';
-
   return t('@bundle request was already @action', array(
     '@bundle' => ucfirst($wrapper->getBundle()),
-    '@action' => $message_action,
+    '@action' => ($state == 'deleted') ? ' rejected' : 'approved',
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/og/c4m_og/c4m_og.module]: add a comment

**Pull request merged**